File: implement-queue-using-stacks/MyQueue232.java
// https://leetcode.com/problems/implement-queue-using-stacks/

import java.util.Stack;

class MyQueue232 {

    Stack<Integer> st1;
    Stack<Integer> st2;

    public MyQueue232() {
        st1 = new Stack<>();
        st2 = new Stack<>();
    }

    // T: O(n), S: O(n)
    public void push(int x) {
        st1.push(x);
    }

    // T: O(1), S: O(1)
    public int pop() {
        if (st2.isEmpty()) {
            while (!st1.isEmpty()) {
                st2.push(st1.pop());
            }
        }
        return st2.pop();
    }

    // T: O(1), S: O(1)
    public int peek() {
        if (st2.isEmpty()) {
            while (!st1.isEmpty()) {
                st2.push(st1.pop());
            }
        }
        return st2.peek();
    }

    // T: O(1), S: O(1)
    public boolean empty() {
        return st1.isEmpty() && st2.isEmpty();
    }

    public static void main(String[] args) {
        var mq = new MyQueue232();
        mq.push(1);
        mq.push(2);
        System.out.println(mq.peek()); // 1
        System.out.println(mq.pop()); // 1
        System.out.println(mq.empty()); // false
    }

}

File: implement-queue-using-stacks/my_queue232.py
# https://leetcode.com/problems/implement-queue-using-stacks/

class MyQueue232:

    def __init__(self):
        self.st1 = []
        self.st2 = []

    # T: O(1), S: O(1)
    def push(self, x: int) -> None:
        self.st1.append(x)

    # T: O(n), S: O(1)
    def pop(self) -> int:
        if not self.st2:
            while self.st1:
                self.st2.append(self.st1.pop())
        return self.st2.pop()

    # T: O(n), S: O(1)
    def peek(self) -> int:
        if not self.st2:
            while self.st1:
                self.st2.append(self.st1.pop())
        return self.st2[-1]

    # T: O(1), S: O(1)
    def empty(self) -> bool:
        return not self.st1 and not self.st2


def main() -> None:
    mq = MyQueue232()
    mq.push(1)
    mq.push(2)
    print(mq.peek())  # 1
    print(mq.pop())   # 1
    print(mq.empty())  # False


if __name__ == "__main__":
    main()
